src/main/java/org/openstreetmap/atlas/utilities/function/TernaryFunction.java

```diff
@@ -0,0 +1,55 @@
+package org.openstreetmap.atlas.utilities.function;
+
+import java.util.Objects;
+import java.util.function.Function;
+
+/**
+ * Represents a function that accepts four arguments and produces a result. This is the four-arity
+ * specialization of {@link Function}.
+ *
+ * @param <S>
+ *            the type of the first argument to the function
+ * @param <T>
+ *            the type of the second argument to the function
+ * @param <U>
+ *            the type of the third argument to the function
+ * @param <R>
+ *            the type of the result of the function
+ * @author lcram
+ */
+@FunctionalInterface
+public interface TernaryFunction<S, T, U, R>
+{
+    /**
+     * Returns a composed function that first applies this function to its input, and then applies
+     * the {@code after} function to the result. If evaluation of either function throws an
+     * exception, it is relayed to the caller of the composed function.
+     *
+     * @param <W>
+     *            the type of output of the {@code after} function, and of the composed function
+     * @param after
+     *            the function to apply after this function is applied
+     * @return a composed function that first applies this function and then applies the
+     *         {@code after} function
+     * @throws NullPointerException
+     *             if after is null
+     */
+    default <W> TernaryFunction<S, T, U, W> andThen(final Function<? super R, ? extends W> after)
+    {
+        Objects.requireNonNull(after);
+        return (final S s, final T t, final U u) -> after.apply(apply(s, t, u));
+    }
+
+    /**
+     * Applies this function to the given arguments.
+     *
+     * @param s
+     *            the first function argument
+     * @param t
+     *            the second function argument
+     * @param u
+     *            the third function argument
+     * @return the function result
+     */
+    R apply(S s, T t, U u);
```

Comment:
Suggestion: Consider adding throws in the signature for apply so that implementers don't have to handle exceptions explicitly. This is the unfortunate downside of Java lambdas, where none of the abstract methods in the functional package `throws` anything which forces you to write `try`...`catch` re-throw - this foils the developers ability to use the lambda shortcuts like `x -> whatever()` or `this::whatever`

Reply:
What should it throw? Just `Exception`?

Reply:
Yes, that is what I would suggest.

Reply:
If I add the `throws Exception`, is there a way to prevent compilation errors in cases like:

```java
static final TernaryOperator<Long> diffBasedLongMerger = 
(before, afterLeft, afterRight) ->
    {
     .
     .
     .
    }
diffBasedLongMerger.apply(before, after1, after2));
```

I get a compiler error on the `diffBasedLongMerger.apply(before, after1, after2));` since it does not catch the exception.

Reply:
Just wrote a quick test case to demo the implications on usability and readability of FunctionalInterfaces that chose to throw CheckedExceptions as a part of their contract or otherwise.

Compare the easierToRead test cases with "messy" test cases.

```java
import org.junit.Assert;
import org.junit.Test;

import java.io.IOException;

/**
 * @author Yazad Khambata
 */
public class ExceptionalFunctionalTest {

    public static final String STR = "Hello";

    @Test
    public void testEasierToRead1() {
        final String s = SomeClass.higerOrderFunction1(this::veryComplexWork, STR);
        verify(s);
    }

    @Test
    public void testEasierToRead2() {
        final String s = SomeClass.higerOrderFunction1(str -> this.veryComplexWork(str), STR);
        verify(s);
    }

    @Test
    public void testEasierToRead3() {
        final ExceptionalFunctional<String, String> ef = this::veryComplexWork;
        final String s = ef.applyQuietly(STR);
        verify(s);
    }

    @Test
    public void testMessy1() {
        final String s = SomeClass.higerOrderFunction2(x -> {
            try {
                return this.veryComplexWork(x);
            } catch (Exception e) {
                throw new RuntimeException(e);
            }
        }, STR);
        verify(s);
    }

    @Test
    public void testMessy2() {
        ConventionalFunctional<String, String> ef = x -> {
            try {
                return this.veryComplexWork(x);
            } catch (final Exception e) {
                throw new RuntimeException(e);
            }
        };
        final String s = ef.apply(STR);
        verify(s);
    }

    private String veryComplexWork(final String str) throws Exception {
        if (false) {
            throw new IOException("ha!");
        }

        return str.toUpperCase();
    }

    private void verify(final String s) {
        Assert.assertEquals("HELLO", s);
    }

    static class SomeClass {
        public static <X, Y>  X higerOrderFunction1(final ExceptionalFunctional<X, Y> exceptionalFunctional, final Y y) {
            return exceptionalFunctional.applyQuietly(y);
        }

        public static <X, Y>  X higerOrderFunction2(final ConventionalFunctional<X, Y> exceptionalFunctional, final Y y) {
            return exceptionalFunctional.apply(y);
        }
    }

    @FunctionalInterface
    interface ExceptionalFunctional<X, Y> {
        X apply(Y y) throws Exception;

        default X applyQuietly(final Y y) {
            try {
                return this.apply(y);
            } catch (final Exception e) {
                throw new IllegalStateException(e);
            }
        }
    }

    @FunctionalInterface
    interface ConventionalFunctional<X, Y> {
        X apply(Y y);
    }
}
```